src/aio/search.rs

```diff
@@ -95,81 +107,147 @@ impl SearchFuture {
     }
 
     // Process a control response to extract the control URL
-    fn handle_control_resp(addr: SocketAddr, resp: Bytes) -> Result<String, SearchError> {
+    fn handle_control_resp(addr: SocketAddr, resp: Bytes) -> Result<(String, String), SearchError> {
         debug!("handling control response from: {}", addr);
 
         // Create a cursor over the response data
         let c = std::io::Cursor::new(&resp);
 
         // Parse control URL out of body
-        let url = parsing::parse_control_url(c)?;
+        let urls = parsing::parse_control_urls(c)?;
+
+        Ok(urls)
+    }
+
+    fn request_schemas(
```

Comment:
Do you think that using async would make the code simpler and cleaner here? I have no issue whatsoever with dropping the raw future implementations and using async functions for clarity.

Reply:
Yes, definitely. I just wanted to preserve as much of your code as possible, so as not to rewrite the whole story. But if you are okay with it, I will try a more `async` approach, without the `Future` implementation.